### ISSUE TITLE
Progress on manual focus?

### ISSUE THREAD

**Issue by leoshmu** [0015/ESP32-OV5640-AF#9]
Thank you for this wonderful project. I'm wondering if there are particular challenges with enabling manual focus. It should consist of sending a command via i2c to the proper register right? Are there registers for setting the position of the lens that are known?

**Comment by leoshmu** [comment]
Looking a little more closely at this, I see you are using the 0x3022 register to get the autofocus started, which is what the arducam ov5640 lists as the register for starting autofocus
However, I found this: https://cdn.sparkfun.com/datasheets/Sensors/LightImaging/OV5640_datasheet.pdf

And in they suggested in section 7-11 that 0x3602 is a register with default value of 0x50 that controls the VCM position. Do you think this could work to allow manual control of the vcm position?

**Comment by 0015** [comment]
@leoshmu As I mentioned in my channel comment, I tried to create firmware for manual focus control, but I couldn't produce the desired result. No updates have been made since then. If you modify and test my code and make a pull request, I will test it myself.

**Comment by leoshmu** [comment]
Thanks so much, I will do my best. If you wouldn't mind sharing what you tried even though it didn't work that would be much appreciated. If you don't have time I understand

**Comment by mfauss** [comment]
For anyone wishing to just manually set the VCM for focus, you can write to the VCM control directly.
The auto focus feature from this repo downloads firmware for performing the auto focus operation to the MCU on the OV5640. - My guess is that firmware was supplied originally by OminVision support.  To manually set the focus, you will not want the autofocus running on the MCU as it will be in conflict as it attempts to control the VCM itself.  There may be an interface into the firmware to set the focus manually, but I am not aware of it.

Here is an example of code I tested on the XIAO ESP32S3 and OV5640 from Seeed.

The registers used below are from the OV5640 datasheet.  If you have another board or library, you may need to use a different function for writing the registers over I2C to the camera.

constexpr int REG_VCM_CONTROL_0 = 0x3602;
constexpr int REG_VCM_CONTROL_1 = 0x3603;
uint16_t focus = 0;
int set_focus(sensor_t *s, uint16_t vcm_value)
{
  // Set DAC bits 3:0 into reg bits 7:4
  s->set_reg(s, REG_VCM_CONTROL_0, 0xF0, vcm_value << 4);
  // Set DAC bits 9:4 into reg bits 5:0 
  s->set_reg(s, REG_VCM_CONTROL_1, 0x3F, vcm_value >> 4);
  focus = vcm_value;
  Serial.println((std::string("Focus set to ") + std::to_string(focus)).c_str());
  return 0;
}

**Comment by abdalrahimnaser** [comment]
Hey @mfauss - Thanks a lot! that saved me quite a lot of time. I would suggest @0015 to merge that in the main driver code.

**Comment by 0015** [comment]
@abdalrahimnaser @mfauss Yes, if anyone makes a PR, I'll merge it into the main branch. Thank you.
